File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/pedropathing/PedroConstants.java
package org.firstinspires.ftc.teamcode.pedropathing;

import com.pedropathing.follower.Follower;
import com.pedropathing.follower.FollowerConstants;
import com.pedropathing.ftc.FollowerBuilder;
import com.pedropathing.ftc.drivetrains.MecanumConstants;
import com.pedropathing.ftc.localization.Encoder;
import com.pedropathing.ftc.localization.constants.DriveEncoderConstants;
import com.pedropathing.paths.PathConstraints;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;

public class PedroConstants {
    public static FollowerConstants followerConstants = new FollowerConstants()
            .mass(11.5)
            .forwardZeroPowerAcceleration(-42.2235)
            .lateralZeroPowerAcceleration(-129.9665);

    public static MecanumConstants driveConstants = new MecanumConstants()
            .maxPower(1)
            .rightFrontMotorName("FR")
            .rightRearMotorName("BR")
            .leftRearMotorName("BL")
            .leftFrontMotorName("FL")
            .leftFrontMotorDirection(DcMotorSimple.Direction.REVERSE)
            .rightRearMotorDirection(DcMotorSimple.Direction.FORWARD)
            .rightFrontMotorDirection(DcMotorSimple.Direction.FORWARD)
            .leftRearMotorDirection(DcMotorSimple.Direction.REVERSE)
            .xVelocity(73.53)
            .yVelocity(57.88);

    public static DriveEncoderConstants localizerConstants = new DriveEncoderConstants()
            .rightFrontMotorName("FR")
            .rightRearMotorName("BR")
            .leftRearMotorName("BL")
            .leftFrontMotorName("FL")
            .leftFrontEncoderDirection(Encoder.REVERSE)
            .rightRearEncoderDirection(Encoder.FORWARD)
            .rightFrontEncoderDirection(Encoder.FORWARD)
            .leftRearEncoderDirection(Encoder.REVERSE)
            .robotLength(13)
            .robotWidth(16.5)
            .forwardTicksToInches(0.0070797)
            .strafeTicksToInches(0.006324)
            .turnTicksToInches(0.01214);

    public static PathConstraints pathConstraints = new PathConstraints(0.99, 100, 1, 1);

    public static Follower createFollower(HardwareMap hardwareMap) {
        return new FollowerBuilder(followerConstants, hardwareMap)
                .driveEncoderLocalizer(localizerConstants)
                .pathConstraints(pathConstraints)
                .mecanumDrivetrain(driveConstants)
                .build();
    }
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/pedropathing/PedroConstants.kt
package org.firstinspires.ftc.teamcode.pedropathing

import com.pedropathing.control.FilteredPIDFCoefficients
import com.pedropathing.control.PIDFCoefficients
import com.pedropathing.follower.Follower
import com.pedropathing.follower.FollowerConstants
import com.pedropathing.ftc.FollowerBuilder
import com.pedropathing.ftc.drivetrains.MecanumConstants
import com.pedropathing.paths.PathConstraints
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.subsystems.mecanum.MecanumLocalizer

object PedroConstants {
    // Technical constants
    val driveConstants: MecanumConstants = MecanumConstants().apply {
        maxPower(1.0)
        leftFrontMotorName ="FL"
        rightRearMotorName = "BR"
        rightFrontMotorName = "FR"
        leftRearMotorName = "BL"
        leftFrontMotorDirection = DcMotorSimple.Direction.REVERSE
        rightRearMotorDirection = DcMotorSimple.Direction.FORWARD
        rightFrontMotorDirection = DcMotorSimple.Direction.FORWARD
        leftRearMotorDirection = DcMotorSimple.Direction.REVERSE
    }

    val localizer = MecanumLocalizer()

    // Following tuning specific constants
    var followerConstants: FollowerConstants = FollowerConstants().apply {
        mass = 5.0
        translationalPIDFCoefficients(PIDFCoefficients(0.1, 0.0, 0.01, 0.0))
        headingPIDFCoefficients(PIDFCoefficients(0.1, 0.0, 0.01, 0.0))
        drivePIDFCoefficients(FilteredPIDFCoefficients(0.1, 0.0, 0.01, 0.6, 0.0))
        centripetalScaling = 0.005
    }

    var pathConstraints: PathConstraints = PathConstraints(0.99, 100.0, 1.0, 1.0)

    @JvmStatic
    fun createFollower(hardwareMap: HardwareMap): Follower {
        return FollowerBuilder(followerConstants, hardwareMap).apply {
            setLocalizer(localizer)
            pathConstraints(pathConstraints)
            mecanumDrivetrain(driveConstants)
        }.build()
    }
}